service/src/main/java/bio/terra/tanagra/service/artifact/CohortService.java

```diff
@@ -127,6 +144,87 @@ public Cohort updateCohort(
+      String userEmail,
+      @Nullable String displayName,
+      @Nullable String description) {
+    Cohort original = getCohort(studyId, cohortId);
+
+    // CohortRevision
+    Map<String, CohortRevision> clonedRevisionMap = new HashMap<>(); // to create reviews
+    List<CohortRevision> clonedRevisionList = new ArrayList<>(); // to maintain original order
+    original
+        .getRevisions()
+        .forEach(
+            revision -> {
+              CohortRevision clonedRevision =
+                  CohortRevision.builder()
+                      .version(revision.getVersion())
+                      .setIsMostRecent(revision.isMostRecent())
+                      .setIsEditable(revision.isEditable())
+                      .createdBy(userEmail)
+                      .lastModifiedBy(userEmail)
+                      .recordsCount(revision.getRecordsCount())
+                      // CriteriaGroupSections (shallow copy): they are written to DB & fetched.
+                      // Any ids are used in conjunction with cohort_revision_id as primary key
+                      .sections(revision.getSections())
+                      .build();
+              clonedRevisionMap.put(original.getId(), clonedRevision);
+              clonedRevisionList.add(clonedRevision);
+            });
+
+    // Cohort
+    Cohort.Builder cohortBuilder =
+        Cohort.builder()
+            .underlay(original.getUnderlay())
+            .displayName(
+                displayName != null ? displayName : "Copy of: " + original.getDisplayName())
+            .description(
+                description != null ? description : "Copy of: " + original.getDescription())
+            .createdBy(userEmail)
+            .lastModifiedBy(userEmail)
+            .revisions(clonedRevisionList);
+
+    // CohortReview
+    ResourceCollection authorizedReviewIds =
+        accessControlService.listAuthorizedResources(
+            SpringAuthentication.getCurrentUser(),
+            Permissions.forActions(REVIEW, READ),
+            ResourceId.forCohort(studyId, cohortId),
+            0,
+            MAX_REVIEW_SIZE);
+    List<Review> clonedReviewList = new ArrayList<>();
+    reviewService
+        .listReviews(authorizedReviewIds, 0, MAX_REVIEW_SIZE)
+        .forEach(
+            review ->
+                clonedReviewList.add(
+                    Review.builder()
+                        .displayName("Copy of: " + review.getDisplayName())
+                        .description("Copy of: " + review.getDescription())
+                        .size(review.getSize())
+                        .revision(clonedRevisionMap.get(review.getRevision().getId()))
+                        .createdBy(userEmail)
+                        .lastModifiedBy(userEmail)
+                        .isDeleted(review.isDeleted())
+                        .build()));
+
+    // Annotations
+
+
+
+    // TODO: pending clone of annotations
+    cohortDao.createCohort(cohortId, cohortBuilder.build());
```

Comment:
It might be easier to do this all inside the Dao classes. e.g. `cohortDao.cloneCohort(originalCohortId, clonedCohortId, clonedCohortDisplayName, clonedCohortDescription)`. Then you can do it all in a single DB transaction, and just copy the table rows instead of worrying about all the objects.

Reply:
Yes, I plan to wrap it in a single function

Reply:
Re-read, and i got your comment now. 
I will talk to you tomorrow, here is my thought - copying all the DB rows for all tables is definitely easier, but I am wondering if this could copy data issues (if any). thought of building objects and then writing DB entries in a single transaction
not a priority - if we want to selectively copy either based on revision, or some other constraint this might help.